wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/monitoring/ClusterTopologyMonitorImpl.java

```diff
@@ -0,0 +1,851 @@
+        && System.nanoTime() < this.ignoreNewFailoverRequestsEndTimeNano.get()) {
+
+      // Previous failover has just completed. We can use results of it without triggering a new failover.
+      List<HostSpec> currentHosts = this.topologyMap.get(this.clusterId);
+      if (currentHosts != null) {
+        return currentHosts;
+      }
+    }
+
+    if (writerImportant) {
+      final Connection monitoringConnection = this.monitoringConnection.get();
+      this.monitoringConnection.set(null);
+      this.isVerifiedWriterConnection = false;
+      this.closeConnection(monitoringConnection, true);
+    }
+
+    return this.waitTillTopologyGetUpdated(timeoutMs);
+  }
+
+  @Override
+  public List<HostSpec> forceRefresh(@Nullable Connection connection, final long timeoutMs)
+      throws SQLException, TimeoutException {
+
+    if (this.isVerifiedWriterConnection) {
+      //push monitoring thread to refresh topology with a verified connection
+      return this.waitTillTopologyGetUpdated(timeoutMs);
+    }
+
+    // otherwise use provided unverified connection to update topology
+    return this.fetchTopologyAndUpdateCache(connection);
+  }
+
+  protected List<HostSpec> waitTillTopologyGetUpdated(final long timeoutMs) throws TimeoutException {
+
+    List<HostSpec> currentHosts = this.topologyMap.get(this.clusterId);
+    List<HostSpec> latestHosts;
+
+    synchronized (this.requestToUpdateTopology) {
+      this.requestToUpdateTopology.set(true);
+
+      // Notify monitoring thread (that might be sleeping) that topology should be refreshed immediately.
+      this.requestToUpdateTopology.notifyAll();
+    }
+
+    if (timeoutMs == 0) {
+      return currentHosts;
+    }
+
+    final long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs);
+    while (currentHosts == (latestHosts = this.topologyMap.get(this.clusterId))
+        && System.nanoTime() < end) {
+      try {
+        synchronized (this.topologyUpdated) {
+          this.topologyUpdated.wait(1000);
+        }
+      } catch (InterruptedException ex) {
+        Thread.currentThread().interrupt();
+        return null;
+      }
+    }
+
+    if (System.nanoTime() >= end) {
+      throw new TimeoutException(Messages.get(
+          "ClusterTopologyMonitorImpl.topologyNotUpdated",
+          new Object[]{timeoutMs}));
+    }
+
+    return latestHosts;
+  }
+
+  @Override
+  public void close() throws Exception {
+    this.stop.set(true);
+
+    // It breaks a waiting/sleeping cycles in monitoring thread
+    synchronized (this.requestToUpdateTopology) {
+      this.requestToUpdateTopology.set(true);
+      this.requestToUpdateTopology.notifyAll();
+    }
+
+    // Waiting for 30s gives a thread enough time to exit monitoring loop and close database connection.
+    if (!this.monitorExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+      this.monitorExecutor.shutdownNow();
+    }
+  }
+
+  @Override
+  public void run() {
```

Comment:
Agree. It might be challenging to read the file. I created a new ticket for that. https://github.com/orgs/awslabs/projects/91/views/1?pane=issue&itemId=79754055